foundations/src/telemetry/log/internal.rs

```diff
@@ -1,11 +1,67 @@
 use super::init::LogHarness;
 use crate::telemetry::scope::Scope;
 use slog::{Logger, OwnedKV, SendSyncRefUnwindSafeKV};
+use std::ops::Deref;
 use std::sync::Arc;
 
+pub const MAX_LOG_GENERATION: u32 = 1000;
+pub const EXCEEDED_MAX_LOG_GENERATION_ERROR: &str = "foundations::telemetry::log: maximum logger generation exceeded (are add_fields! or set_verbosity being called in a loop?)";
+
+// The slog_logger() function exposes Arc<RwLock<Logger>> as a public
+// interface, so we can't store the generation number where it would make the
+// most sense to store it, which would be inside the innermost RwLock, right
+// next to the Logger. So we need two nested levels of Arcs and RwLocks.
+//
 // NOTE: we intentionally use a lock without poisoning here to not
 // panic the threads if they just share telemetry with failed thread.
-pub(crate) type SharedLog = Arc<parking_lot::RwLock<Logger>>;
+pub(crate) type SharedLog = Arc<parking_lot::RwLock<LogContext>>;
```

Comment:
The comment above explains why we cannot get away with just a single `Arc` and a single `RwLock`, with both the `Logger` and the `u32` living underneath the `RwLock`.

If we really wanted to, we could get rid of the outer `RwLock`, and just have `SharedLog` be defined as an `Arc<LogContext>`. That would mean we would need interior mutability inside `LogContext` for updating the generation number; so that could be changed to an `AtomicU32`.

I initially did it that way; but I found myself writing voluminous comments everywhere in order to justify/excuse what I was doing. It felt like a code smell, to me. So I added the extra `RwLock` layer. I feel like this way of doing it is more readily comprehensible to people who might be reading the code.

Reply:
Personally, I think it would be better to use an `AtomicU32` on `generation`. It's weird to me to have a struct that has only partial interior mutability, as it forces what you did here - two layers of locks. Since we're forced into the `Arc<RwLock<Logger>>` for storing the logger itself, I think the next best choice is making the other field(s) on `LogContext` have their own locks.

One benefit is that it would reduce work on the hot path of actually writing logs, while it would probably be a similar amount of work (not sure how locking a RwLock compares to incrementing + reading an atomic) on the `add_fields!` path.

Reply:
I think this is an improvement. As you note, it's not mission critical that generation is exactly in sync - we don't care if it triggers at 1005 and not 1000, really. Maybe we could even used `Ordering::Relaxed`, but I'm no expert on what exactly that does and whether it's safe. I really like that we don't have an extra lock on the hot path of logging now. `overwrite_from` is a little gross, but I think it's worth it.

Reply:
OK! I'm glad we did this, rather than just me making assumptions about what would be best! :-)

Reply:
With the current implementation, the `generation` field is indeed accessed always under the logger lock, which creates the necessary `happens-before` relationship. Those atomic accesses could be replaced by relaxed operations without  any impact.

As for what those mean, basically it's how strong of a memory barrier the compiler needs to issue depending on the architecture of the processor. For e.g., with strong consistency you don't really get any more expensive than with relaxed in x86_64.  But for ARM, strong consistency is more expensive than relaxed because the memory model of the processor is weaker than x86_64.

As for atomic increments vs lock acquisition, the former may be cheaper in some architectures (e.g. x86_64 has fetch_and_inc which is commutative and much more scalable in terms of concurrency than compare_and_swap, with the latter being the primitive used for locks).